simulator/tests/randpa_finality_tests.cpp

```diff
@@ -28,6 +28,295 @@ TEST(randpa_finality, two_full_nodes) {
+    g[4] = {{5, random()}, {6, random()}};
+    g[6] = {{7, random()}};
+
+    runner.load_graph(g);
+    runner.add_stop_task(9 * runner.get_slot_ms());
+    runner.run<RandpaNode>();
+    for (auto i = 0; i<nodes_amount; ++i) {
+        if (g[i] != empty_vector) {
+            EXPECT_EQ(get_block_height(runner.get_db(i).last_irreversible_block_id()), 7);
+        }
+    }
+}
+
+TEST(randpa_finality, bp_over_vertical_lines_fn) {
+    size_t nodes_amount = 42;
+    auto runner = TestRunner(nodes_amount);
+    vector<node_type> nodetypes(nodes_amount, node_type::FN);
+    nodetypes[0] = node_type::BP;
+    nodetypes[41] = node_type::BP;
+    runner.load_nodetypes(nodetypes);
+
+    auto max_delay = 40;
+    auto min_delay = 10;
+    auto random = [&](){ return (rand() % (max_delay - min_delay)) + min_delay; };
+
+    graph_type g;
+    vector<pair<int, int> > empty_vector;
+    vector<pair<int, int>> temp_pairs;
+    for (auto i = 1; i<11; ++i) {
+        temp_pairs.push_back({i, random()});
+    }
+    g.push_back(temp_pairs);
+    for (auto i = 1; i<4; ++i) {
+        for (auto j = 1; j<11; ++j) {
+            vector<pair<int, int>> pairs;
+            for (auto k = 1; k<11; ++k) {
+                pairs.push_back({i*10+k, random()});
+            }
+            g.push_back(pairs);
+        }
+    }
+    for (auto i = 0; i<10; ++i) {
+        g.push_back({{41, random()}});
+    }
+    g.push_back(empty_vector);
+
+    runner.load_graph(g);
+    runner.add_stop_task(15 * runner.get_slot_ms());
+    runner.run<RandpaNode>();
+    for (auto i = 0; i<nodes_amount; ++i) {
+        if (g[i] != empty_vector) {
+            EXPECT_EQ(get_block_height(runner.get_db(i).last_irreversible_block_id()), 13);
+        }
+    }
+}
+
+TEST(randpa_finality, bp_over_horizontal_lines_fn) {
+    size_t nodes_amount = 42;
+    auto runner = TestRunner(nodes_amount);
+    vector<node_type> nodetypes(nodes_amount, node_type::FN);
+    nodetypes[0] = node_type::BP;
+    nodetypes[41] = node_type::BP;
+    runner.load_nodetypes(nodetypes);
+
+    auto max_delay = 40;
+    auto min_delay = 10;
+    auto random = [&](){ return (rand() % (max_delay - min_delay)) + min_delay; };
+
+    graph_type g;
+    vector<pair<int, int>> empty_vector;
+    g.push_back({{1, random()}, {11, random()}, {21, random()}, {31, random()}});
+    for (auto i = 1; i<4; ++i) {
+        for (auto j = 1; j<11; ++j) {
+            vector<pair<int, int>> pairs;
+            pairs.push_back({(j == 10 ? 41 : (i-1)*10 + j + 1), random()});
+            for (auto k = 1; k<11; ++k) {
+                pairs.push_back({i*10+k, random()});
+            }
+            g.push_back(pairs);
+        }
+    }
+    for(auto i = 31; i<nodes_amount - 1; ++i)
+        g.push_back({{i+1, random()}});
+    g.push_back(empty_vector);
+
+    runner.load_graph(g);
+    runner.add_stop_task(15 * runner.get_slot_ms());
+    runner.run<RandpaNode>();
+    for (auto i = 0; i<nodes_amount; ++i) {
```

Comment:
spaces around operations
`i < nodes_amount`

Reply:
Added braces {}, assert, more readable listing of node types, deleted test without BP (randpa_finality.two_full_nodes), added spaces to operators.

Reply:
Test randpa_finality.fullnodes_over_round_ring does not pass after adding 
`    auto seed = time(NULL);
    std::srand(seed);
    std::cout << "Random number generator seeded to " << time(NULL) << std::endl;
    std::cout << seed;
`
in DAObet/simulator/main.cpp.
All the time testcase has different `get_block_height(runner.get_db(i).last_irreversible_block_id())` for nodes.